### ISSUE TITLE
Support final LR annealing phase

### ISSUE THREAD

**Issue by ddobbelaere** [glinscott/nnue-pytorch#37]
See also https://github.com/official-stockfish/Stockfish/issues/3274

The optimizer Ranger states the following in its documentation:

> Best training results - use a 75% flat lr, then step down and run lower lr for 25%, or cosine descend last 25%.
>
> Per extensive testing - It's important to note that simply running one learning rate the entire time will not produce optimal results.
> Effectively Ranger will end up 'hovering' around the optimal zone, but can't descend into it unless it has some additional run time at a lower rate to drop down into the optimal valley.

Adding support for this to the trainer seems very interesting. See https://github.com/lessw2020/Ranger-Mish-ImageWoof-5/blob/b0aa73508870de072329d058f0add165da462d6d/train.py#L56 for a sample implementation.

**Comment by vondele** [comment]
there is a branch now https://github.com/glinscott/nnue-pytorch/tree/learn_sched that implements a simple variant. 

I see improved performance with this branch.. also used a smaller batch size.

**Comment by ddobbelaere** [comment]
Just noticed pytorch has an LR scheduler that accepts a lambda function for LR as function of epoch: https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.LambdaLR

Might be interesting to add some final cosine annealing into the mix without much hassle...

**Comment by ddobbelaere** [comment]
I'm sure a lot of things have been tried out by now :)
Currently an exponentially decreasing LR seems optimal...

**Comment by Sopel97** [comment]
not so much, still a lot of stuff left to explore with hyperparameters. I don't think anyone actually tried annealing LRs yet.

**Comment by ddobbelaere** [comment]
> not so much, still a lot of stuff left to explore with hyperparameters. I don't think anyone actually tried annealing LRs yet.

Hmm, ok. Well, in any case, it more like "something left to try" than an issue I think.
